### PR TITLE
A new iterator in Hashtbl to allow safe updates during the traversal

### PR DESCRIPTION
**Pull request by alainfrisch**
It is currently not specified whether the callback passed to [Hashtbl.{iter,fold}] is allowed to modify the hash table.  Looking at the implementation, it is ok to modify the current binding (remove it, or replace it with a different value -- assuming there is a single value that the key), and this is a useful case to preserve.  Other modifications don't break everything, but it cannot easily be predicted whether they can be "seen" during the iteration or not. To allow for more efficient implementations of hash tables (see #328), I propose to introduce a new iterator that covers the useful cases of modifying the current binding (already with greatly improved performances) and explicitly saying that any other modification to the hash table during the traversal is not supported.

### CODE REVIEW

File: stdlib/moreLabels.mli [hunk 1]
```diff
@@ -34,6 +34,7 @@ module Hashtbl : sig
   val remove : ('a, 'b) t -> 'a -> unit
   val replace : ('a, 'b) t -> key:'a -> data:'b -> unit
   val iter : f:(key:'a -> data:'b -> unit) -> ('a, 'b) t -> unit
+  val update: f:(key:'a -> date:'b -> 'b option) -> ('a, 'b) t -> unit
```
Comment (Drup) [inline stdlib/moreLabels.mli]: `data` ?

File: stdlib/hashtbl.mli [hunk 2]
```diff
@@ -120,7 +120,19 @@ val iter : ('a -> 'b -> unit) -> ('a, 'b) t -> unit
...
+   by [f] during the iteration.
+*)
+
+val update: ('a -> 'b -> 'b option) -> ('a, 'b) t -> unit
+(** [Hashtbl.update f tbl] applies [f] to all bindings in table [tbl]
+    and update each binding depending on the result of [f].  If [f]
+    returns [None], the binding is discared.  If it returns [Some
+    new_val], the binding is update to associare the key to [new_val].
+
+    Other comments for {!Hashtbl.update} apply as well.  *)
```
Comment (Drup) [inline stdlib/hashtbl.mli]: What does this sentence mean ?

File: stdlib/moreLabels.mli (hunk 1 above)
Comment (alainfrisch) [inline stdlib/moreLabels.mli]: Thanks!

File: stdlib/hashtbl.mli (hunk 2 above)
Comment (alainfrisch) [inline stdlib/hashtbl.mli]: I meant Hashtbl.iter.  This is to avoid copying most of the doc text from Hashtbl.iter.

File: stdlib/hashtbl.mli [hunk 3]
```diff
@@ -120,7 +120,19 @@ val iter : ('a -> 'b -> unit) -> ('a, 'b) t -> unit
...
+   is entirely random.
+
+   The behavior is not defined if the hash table is modified
+   by [f] during the iteration.
+*)
+
+val update: ('a -> 'b -> 'b option) -> ('a, 'b) t -> unit
+(** [Hashtbl.update f tbl] applies [f] to all bindings in table [tbl]
+    and update each binding depending on the result of [f].  If [f]
+    returns [None], the binding is discared.  If it returns [Some
```
Comment (thrasibule) [inline stdlib/hashtbl.mli]: discarded

File: stdlib/hashtbl.mli [hunk 4]
```diff
@@ -120,7 +120,19 @@ val iter : ('a -> 'b -> unit) -> ('a, 'b) t -> unit
...
+    new_val], the binding is update to associare the key to [new_val].
```
Comment (thrasibule) [inline stdlib/hashtbl.mli]: updated to associate

File: stdlib/hashtbl.mli (hunk 3 above)
Comment (alainfrisch) [inline stdlib/hashtbl.mli]: Thanks!  (The number of typos is just embarrassing...)

**Comment by gasche** [comment]
I am a bit worried by this proposal.

First, I don't think naming this function is simple, and I suspect that `update` is a bad name for it (if I was asked to guess a type for it, it would be `('a, 'b) Hashtbl.t -> 'a -> 'b -> unit`, and in fact I would guess it is a synonym of `replace`). Given what you seem to suggest as a semantics, I think that `filter_map_inplace` would be a better name; `filter_map` denotes that the filter may return a new element, and `_inplace` denotes the fact that the original structure is modified.

(I just now realize by grepping the sources that `filter_map_inplace` in fact exists as a function in Batteries: [see documentation](http://ocaml-batteries-team.github.io/batteries-included/hdoc2/BatHashtbl.html#6_Traversing). I did not remember it when writing the suggestion above but that still have been an inconscious influence.)

Then, and I think this is more problematic, I am not sure I understand what you suggest and implemented by proposing that we would "explicitly say that any other modification to the hash table during the traversal is not supported.". One thing that I would disagree with is changing the implementation of the other functions in such a way that something that worked reliably before would not work anymore with your new implementation -- that would break code and we must not do that.

Will all Hashtbl-using code that was working reliably in 4.02 still work just fine if #328 and #337 are merged?

**Comment by alainfrisch** [comment]
For the name, I'm fine with your suggestion.

> Will all Hashtbl-using code that was working reliably in 4.02 still work just fine if #328 and #337 are merged?

Well, #337 is just about declaring some uses as invalid, it does not actually break any existing code.

#328 could indeed change the behavior of existing code, but: (i) not for the common cases (removing/changing the current binding), and (ii) the current behavior of other cases is already not documented nor uniform (changes done to other bindings can be seen or not during the current iteration depending).

#337 is a preparation for #328: let's explicitly discourage the uncontrolled use of mutation during iteration, let's provide a safe alternative; this will make it easier to accept #328 later, knowing that it will preserve most common uses of uncontrolled mutations anyway.

**Comment by gasche** [comment]
(I decided to continue discuss visible behavior changes with other functions in the other PR, keeping this one focused on a new proposed function.)

**Comment by Drup** [comment]
I agree with @gasche on the naming issue.

**Comment by damiendoligez** [comment]
OK for merging with the new name.

**Comment by alainfrisch** [comment]
Merged (7dce037bdf15c696f9576d4fbf8c87813360b489).

**Pull request closed without merging**